[gloo] Make it possible for gloo TCPStore to take over an existing socket fd

jmswen opened the pull request:
Summary:
This diff allows the `TCPStore` server associated with a gloo process group to listen on an existing socket already bound to a port.

Without the functionality in this diff, canonical initialization of a gloo `ProcessGroup` is fundamentally racy: 1) ask the OS for a free port by creating a socket bound to port 0, 2) close the socket, 3) attempt to initialize a `TCPStore` server that listens on the previously free port. Of course, the problem is that in between steps 2 and 3, another process on the host may have claimed the port, causing `TCPStore` and overall process group initialization to fail. With this diff, it is now possible for users to completely avoid this race (see unit test for how this can be achieved).

Test Plan:
Added new unit test:
  buck2 test caffe2/test/distributed:store

Differential Revision: D46622317

rohan-varma reviewed (commented):
Sounds good to me. cc @H-Huang @kumpera

jmswen commented:
@rohan-varma Would you mind kicking off tests for the latest change? I fixed the test, which was failing on some older platforms without IPv6 support. Thanks!

H-Huang commented on torch/csrc/distributed/c10d/TCPStore.cpp:
```diff
@@ -850,7 +850,9 @@ std::mutex TCPServer::cache_mutex_{};
   auto startCore = [&opts]() {
-    Socket socket = Socket::listen(opts.port);
+    Socket socket = opts.masterListenFd.has_value()
+        ? Socket::listenFromFd(*opts.masterListenFd, opts.port)
```
from a debuggability standpoint, it might still be useful to log the `<address:port>` from opts here even in the case we are using an pre-existing FD. In the current path, that `<address:port>` gets logged as part of `Socket::listen(opts.port);`

jmswen replied:
Thank you for the review, @H-Huang! Currently I have an info line to log the fd and port at the end of `SocketListenFromFdOp::run()`:
```
  C10D_INFO(
      "The server has taken over the listening socket with fd {} and port {}",
      fd_,
      expected_port_);
```

I can extend that to also include the address, which I believe is what you are referring to. Please correct me if I'm misunderstanding.

H-Huang reviewed (approved):
Thanks for adding this, seems super useful! Approved with some minor comments.

jmswen commented:
@pytorchbot merge

jmswen committed:
[gloo] Make it possible for gloo TCPStore to take over an existing socket fd (#103478)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/103478

This diff allows the `TCPStore` server associated with a gloo process group to listen on an existing socket already bound to a port.

Without the functionality in this diff, canonical initialization of a gloo `ProcessGroup` is fundamentally racy: 1) ask the OS for a free port by creating a socket bound to port 0, 2) close the socket, 3) attempt to initialize a `TCPStore` server that listens on the previously free port. Of course, the problem is that in between steps 2 and 3, another process on the host may have claimed the port, causing `TCPStore` and overall process group initialization to fail. With this diff, it is now possible for users to completely avoid this race (see unit test for how this can be achieved).

Test Plan:
Added new unit test:
  buck2 test caffe2/test/distributed:store

Reviewed By: H-Huang

Differential Revision: D46622317

fbshipit-source-id: 9050cda6a8da289bf1452101e3730bc49de413ee

kumpera commented:
@pytorchmergebot merge -f "internal changes  were removed. "
